File: project/training/train_model.ts
import * as tf from '@tensorflow/tfjs-node';
import * as fs from 'fs';
import * as path from 'path';
import { EyeDiseaseModel } from '../lib/eye-disease-model';

const DATASET_PATH = 'c:/Users/VEDANT/OneDrive/Desktop/archive (2)/Eye_diseases';
const BATCH_SIZE = 32;
const EPOCHS = 50;
const VALIDATION_SPLIT = 0.2;

async function loadDataset() {
  const categories = ['Bulging_Eyes', 'Cataracts', 'Crossed_Eyes', 'Glaucoma', 'Uveitis'];
  const images: tf.Tensor4D[] = [];
  const labels: number[] = [];

  for (let i = 0; i < categories.length; i++) {
    const category = categories[i];
    const categoryPath = path.join(DATASET_PATH, category);
    const files = fs.readdirSync(categoryPath);

    console.log(`Loading ${category} images...`);

    for (const file of files) {
      if (file.endsWith('.jpg') || file.endsWith('.png')) {
        const imagePath = path.join(categoryPath, file);
        const imageBuffer = fs.readFileSync(imagePath);
        const image = await tf.node.decodeImage(imageBuffer);
        const resized = tf.image.resizeBilinear(image as tf.Tensor3D, [224, 224]);
        const normalized = resized.div(255.0);
        const expanded = normalized.expandDims(0);

        images.push(expanded as tf.Tensor4D);
        labels.push(i);

        // Cleanup tensors
        image.dispose();
        resized.dispose();
        normalized.dispose();
      }
    }
  }

  // Concatenate all images and create labels tensor
  const xs = tf.concat(images, 0);
  const ys = tf.oneHot(labels, categories.length);

  return { xs, ys };
}

async function trainModel() {
  try {
    console.log('Loading dataset...');
    const { xs, ys } = await loadDataset();

    console.log('Creating model...');
    const model = new EyeDiseaseModel();
    await model.loadModel();

    console.log('Training model...');
    const history = await model.model!.fit(xs, ys, {
      batchSize: BATCH_SIZE,
      epochs: EPOCHS,
      validationSplit: VALIDATION_SPLIT,
      callbacks: {
        onEpochEnd: (epoch, logs) => {
          console.log(`Epoch ${epoch + 1} of ${EPOCHS}`);
          console.log(`Loss: ${logs?.loss.toFixed(4)}, Accuracy: ${logs?.acc.toFixed(4)}`);
        }
      }
    });

    console.log('Training completed!');
    console.log('Final accuracy:', history.history.acc[history.history.acc.length - 1]);

    // Save the model
    await model.model!.save('file://./trained_model');
    console.log('Model saved successfully!');

    // Cleanup
    xs.dispose();
    ys.dispose();
  } catch (error) {
    console.error('Training error:', error);
  }
}

trainModel(); 

File: project/training/train_model.py
import os
import numpy as np
import tensorflow as tf
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Dense, Flatten, Dropout, BatchNormalization
from tensorflow.keras.applications import ResNet50
from tensorflow.keras.optimizers import Adam
import matplotlib.pyplot as plt

# Set random seed for reproducibility
tf.random.set_seed(42)
np.random.seed(42)

# Define constants
IMG_SIZE = 224  # Standard size for ResNet50
BATCH_SIZE = 32
EPOCHS = 50
DATASET_PATH = r"C:\Users\VEDANT\OneDrive\Desktop\archive (2)\Eye_diseases"

# Define classes
CLASSES = ['Bulging_Eyes', 'Cataracts', 'Crossed_Eyes', 'Glaucoma', 'Uveitis']

# Data augmentation for training
train_datagen = ImageDataGenerator(
    rescale=1./255,
    rotation_range=20,
    width_shift_range=0.2,
    height_shift_range=0.2,
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True,
    fill_mode='nearest',
    validation_split=0.2
)

# Load training data
train_generator = train_datagen.flow_from_directory(
    DATASET_PATH,
    target_size=(IMG_SIZE, IMG_SIZE),
    batch_size=BATCH_SIZE,
    class_mode='categorical',
    subset='training'
)

# Load validation data
validation_generator = train_datagen.flow_from_directory(
    DATASET_PATH,
    target_size=(IMG_SIZE, IMG_SIZE),
    batch_size=BATCH_SIZE,
    class_mode='categorical',
    subset='validation'
)

# Create the model
def create_model():
    # Load pre-trained ResNet50 model
    base_model = ResNet50(
        weights='imagenet',
        include_top=False,
        input_shape=(IMG_SIZE, IMG_SIZE, 3)
    )
    
    # Freeze the base model layers
    base_model.trainable = False
    
    # Create the model
    model = Sequential([
        base_model,
        Flatten(),
        Dense(512, activation='relu'),
        BatchNormalization(),
        Dropout(0.5),
        Dense(len(CLASSES), activation='softmax')
    ])
    
    return model

# Create and compile the model
model = create_model()
model.compile(
    optimizer=Adam(learning_rate=0.001),
    loss='categorical_crossentropy',
    metrics=['accuracy']
)

# Train the model
history = model.fit(
    train_generator,
    validation_data=validation_generator,
    epochs=EPOCHS,
    callbacks=[
        tf.keras.callbacks.EarlyStopping(
            monitor='val_loss',
            patience=5,
            restore_best_weights=True
        ),
        tf.keras.callbacks.ReduceLROnPlateau(
            monitor='val_loss',
            factor=0.2,
            patience=3,
            min_lr=0.00001
        )
    ]
)

# Save the model
model.save('eye_disease_model.h5')

# Plot training history
plt.figure(figsize=(12, 4))

# Plot accuracy
plt.subplot(1, 2, 1)
plt.plot(history.history['accuracy'], label='Training Accuracy')
plt.plot(history.history['val_accuracy'], label='Validation Accuracy')
plt.title('Model Accuracy')
plt.xlabel('Epoch')
plt.ylabel('Accuracy')
plt.legend()

# Plot loss
plt.subplot(1, 2, 2)
plt.plot(history.history['loss'], label='Training Loss')
plt.plot(history.history['val_loss'], label='Validation Loss')
plt.title('Model Loss')
plt.xlabel('Epoch')
plt.ylabel('Loss')
plt.legend()

plt.tight_layout()
plt.savefig('training_history.png')
plt.close()

# Print final metrics
print("\nFinal Training Accuracy:", history.history['accuracy'][-1])
print("Final Validation Accuracy:", history.history['val_accuracy'][-1]) 